File: leetcode/Tree/leetcode-124/124.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int max = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        if (root.left == null && root.right == null) {
            return root.val;
        }
        maxPathLocal(root);
        return max;
    }

    private int maxPathLocal(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = maxPathLocal(root.left);
        int right = maxPathLocal(root.right);
        int res = Math.max(root.val, root.val + Math.max(left, right));
        int localMax = Math.max(res, root.val + left + right);
        max = Math.max(max, localMax);
        return res;
    }
}

File: leetcode/Tree/leetcode-124/124.go
package tree

import "math"

// TreeNode Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	if root == nil {
		return 0
	}
	maxSum := math.MinInt32
	maxLocalNode(root, &maxSum)
	return maxSum
}

func maxLocalNode(root *TreeNode, maxSum *int) int {
	if root == nil {
		return 0
	}

	left := maxLocalNode(root.Left, maxSum)
	right := maxLocalNode(root.Right, maxSum)

	res := max(root.Val, root.Val+max(left, right))
	localMax := max(res, root.Val+right+left)
	*maxSum = max(localMax, *maxSum)
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
